MDANSE/Src/MDANSE/Mathematics/Arithmetic.py

```diff
@@ -46,30 +67,85 @@ def get_weights(props, contents, dim):
         for k in list(weights.keys()):
             weights[k] /= np.float64(normFactor)
 
-    return weights, normFactor
-
-
-def weight(props, values, contents, dim, key, symmetric=True, update_partials=False):
-    weights = get_weights(props, contents, dim)[0]
-    weightedSum = None
-    matches = {key % k: k for k in weights}
-
-    for k, val in values.items():
+    weights["sum"] = normFactor
+
+    return weights
+
+
+def assign_weights(
+    values: Dict[str, np.ndarray],
+    weights: Dict[str, float],
+    key: str,
+    symmetric: bool = True,
+):
+    """Updates the scaling factors of partial datasets, without
+    modifying the data.
+
+    Parameters
+    ----------
+    values : Dict[str, np.ndarray]
+        Dictionary of data arrays containing analysis results.
+    weights : Dict[str, float]
+        Dictionary of scaling factors per dataset
+    key : str
+        A string data set name with formatting elements (placeholders for chemical element labels)
+    symmetric : bool, optional
+        do not generate results for the same elements in a different sequence, by default True
+
+    Returns
+    -------
+    np.ndarray
+        total sum of all the component arrays scaled by their weights
+    """
+    matches = {key % k: k for k in weights if k not in ["sum"]}
+    dim = key.count("%s")
+
+    for k in values.keys():
         if k not in matches:
             continue
 
         if symmetric:
             permutations = set(itertools.permutations(matches[k], r=dim))
-            w = sum([weights.pop(p) for p in permutations])
+            w = sum([weights[p] for p in permutations])
```

Comment:
```suggestion
            permutations = itertools.permutations(matches[k], r=dim)
            w = sum(weights[p] for p in permutations)
```

Reply:
The change from 

```
permutations = set(itertools.permutations(matches[k], r=dim))
```

to

```
permutations = itertools.permutations(matches[k], r=dim)
```

causes some test failures because matches[k] could be ("S", "S") which will give ("S", "S") twice which we don't want. I've left in the original version.